### PR TITLE
Snoble/fix bigdecimal semigroup

### PR DESCRIPTION
**Pull request by snoble**
Fixes bug where there was no Semigroup for BigDecimal. With tests.

### CODE REVIEW

File: algebird-core/src/main/scala/com/twitter/algebird/Semigroup.scala [hunk 1]
```diff
@@ -195,6 +195,7 @@ object Semigroup extends GeneratedSemigroupImplicits with ProductSemigroups with
   implicit def jshortSemigroup: Semigroup[JShort] = JShortRing
   implicit def longSemigroup: Semigroup[Long] = LongRing
   implicit def bigIntSemigroup: Semigroup[BigInt] = BigIntRing
+  implicit def bigDecimalSemigroup: Semigroup[BigDecimal] = implicitly[Ring[BigDecimal]]
```
Comment (oscar-stripe) [inline algebird-core/src/main/scala/com/twitter/algebird/Semigroup.scala:198]: can we add this for `Monoid` and `Group` too?

File: algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala [hunk 2]
```diff
@@ -52,6 +52,30 @@ class BatchedLaws extends CheckProperties {
   testBatchedMonoid[String]("String", 10)
   testBatchedMonoid[String]("String", 100)
   testBatchedMonoid[String]("String", 1000000)
+
+  def testBatchedSemigroup[A: Arbitrary: Semigroup](name: String, size: Int): Unit = {
+    implicit val m: Semigroup[Batched[A]] = Batched.compactingSemigroup[A](size)
+    property(s"CountMinSketch[$name] batched at $size is a Semigroup") {
```
Comment (oscar-stripe) [inline algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala]: what is countminsketch got to do with anything here?

File: algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala [hunk 3]
```diff
@@ -52,6 +52,30 @@ class BatchedLaws extends CheckProperties {
   testBatchedMonoid[String]("String", 10)
   testBatchedMonoid[String]("String", 100)
   testBatchedMonoid[String]("String", 1000000)
+
+  def testBatchedSemigroup[A: Arbitrary: Semigroup](name: String, size: Int): Unit = {
```
Comment (oscar-stripe) [inline algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala]: why was this added here. Seems a weird way to test bigdecimal?

Can we instead add a file like `NumericAlgebraTests.scala` that has something like:
```scala
class NumericAlgebraTests extends CheckProperties {
  property("test bigdecimal") {
    // check resolution:
    implicitly[Semigroup[BigDecimal]]
    implicitly[Monoid[BigDecimal]]
    implicitly[Group[BigDecimal]]
    // now test all the laws:
    ringLaws[BigDecimal]
  }
```
and similar for the numeric types we care about?

File: algebird-core/src/main/scala/com/twitter/algebird/Semigroup.scala (hunk 1 above)
Comment (snoble) [reply algebird-core/src/main/scala/com/twitter/algebird/Semigroup.scala:198]: they're already there... this was the one gap

File: algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala (hunk 3 above)
Comment (snoble) [reply algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala]: yeah... I wanted to add _something_ as a test... even if it was crude... I'll add this instead

File: algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala (hunk 2 above)
Comment (snoble) [reply algebird-test/src/test/scala/com/twitter/algebird/BatchedTest.scala]: copy-pasta... I didn't read at all but it comes directly from `def testBatchedMonoid`

File: algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      // check resolution:
+      implicitly[Semigroup[BigDecimal]]
+      implicitly[Monoid[BigDecimal]]
+      implicitly[Group[BigDecimal]]
+      // now test all the laws:
+      ringLaws[A]
+    }
+
+  testRingAndBelowProperties[Int]("int")
+  testRingAndBelowProperties[BigDecimal]("bigdecimal")
```
Comment (snoble) [inline algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala]: This one keeps failing for me with errors like
```
[info] NumericAlgebraTests:
[info] - test int
[info] - test bigdecimal *** FAILED ***
[info]   GeneratorDrivenPropertyCheckFailedException was thrown during property evaluation.
[info]    (CheckProperties.scala:12)
[info]     Falsified after 0 successful property evaluations.
[info]     Location: (CheckProperties.scala:12)
[info]     Occurred when passed generated values (
[info]       arg0 = 1.627209798210513,
[info]       arg1 = 606303697535274687,
[info]       arg2 = -0.003402952952309675094511817686947572
[info]     )
[info]     Label of failing property:
[info]       isDistributiveDifferentTypes
```

Comment (johnynek) [reply algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala]: it may not be distributive

did you check?

Float and Double are not.

Comment (snoble) [reply algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala]: when I test with
```
scala> def foo[T](a: T, b: T, c: T)(implicit rng: Numeric[T]) = {
     |     Equiv[T].equiv(rng.times(a, rng.plus(b, c)), rng.plus(rng.times(a, b), rng.times(a, c))) &&
     |     Equiv[T].equiv(rng.times(rng.plus(b, c), a), rng.plus(rng.times(b, a), rng.times(c, a)))
     | }
```
it seems to work with the failing values as BigDecimal

Comment (snoble) [reply algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala]: ahhh... it's rounding errors
```
scala> val a = BigDecimal(1.627209798210513)
a: scala.math.BigDecimal = 1.627209798210513

scala> val b = BigDecimal("606303697535274687")
b: scala.math.BigDecimal = 606303697535274687

scala> val c = BigDecimal("-0.003402952952309675094511817686947572")
c: scala.math.BigDecimal = -0.003402952952309675094511817686947572
scala> def bar[T](a: T, b: T, c: T)(implicit rng: Numeric[T]) = {
     | (rng.times(a, rng.plus(b, c)), rng.plus(rng.times(a, b), rng.times(a, c)), rng.times(rng.plus(b, c), a), rng.plus(rng.times(b, a), rng.times(c, a)))
     | }

scala> bar(a,b,c)
res9: (scala.math.BigDecimal, scala.math.BigDecimal, scala.math.BigDecimal, scala.math.BigDecimal) = (986583317320662231.5814891877975833,986583317320662231.581489187797583304121164227599838248,986583317320662231.5814891877975833,986583317320662231.581489187797583304121164227599838248)
```

How are those situations dealt with?

Comment (johnynek) [reply algebird-test/src/test/scala/com/twitter/algebird/NumericAlgebraTests.scala]: you can set up a custom `Equiv` instance:

```scala
implicit val closeEnough: Equiv[BigDecimal] = new Equiv[BigDecimal] {
  val eps = BigDecimal(1e-10) // or something.
  def equiv(a: BigDecimal, b: BigDecimal) = 2 * (b - a).abs / (b.abs + a.abs) < eps
}
```

File: algebird-test/src/main/scala/com/twitter/algebird/BaseProperties.scala [hunk 5]
```diff
@@ -27,8 +27,8 @@ import scala.math.Equiv
 object BaseProperties extends MetricProperties {
   val arbReasonableBigDecimals: Arbitrary[BigDecimal] = Arbitrary(
     for {
-      scale <- Gen.choose(-128, +128)
-      base <- implicitly[Arbitrary[BigInt]].arbitrary
+      scale <- Gen.choose(-4, +4)
+      base <- implicitly[Arbitrary[Int]].arbitrary
```
Comment (snoble) [inline algebird-test/src/main/scala/com/twitter/algebird/BaseProperties.scala]: @johnynek: This may not be a good idea. I can go into it but when you add two BigDecimals that have sufficiently different scales it is possible to lose precision. And when testing out the distributive properties of rings the way in which the precision is lost is not distributive. I suspect nobody is surprised by this though so it's reasonable to limit the tests to cases where precision can't be lost.

Comment (johnynek) [reply algebird-test/src/main/scala/com/twitter/algebird/BaseProperties.scala]: This is fine, but can we do the biggest scale we can do with the tests passing consistently? Would 24 work for instance?

**Comment by snoble** [comment]
@johnynek: `(-8,+8)` was as far as I could get and consistently pass

**Comment by snoble** [comment]
turns out I didn't try hard enough to break it :/

**Comment by johnynek** [comment]
👍
